### PR TITLE
[FLINK-33757] Implement restore tests for Rank node

### PR DESCRIPTION
**Pull request by jnh5y**
## What is the purpose of the change

Implement restore tests for Rank node

## Verifying this change

This change added tests and can be verified as follows:

* Added restore tests for Rank node which verifies the generated compiled plan with the saved compiled plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Review by dawidwys — COMMENTED** [review]
The migrated tests look good.

File: flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-rank_1/rank-test/plan/rank-test.json [hunk 1]
```diff
@@ -81,19 +52,20 @@
...
-      "endIndex" : 0
+      "type" : "Constant",
+      "start" : 1,
+      "end" : 1
     },
     "rankStrategy" : {
       "type" : "AppendFast"
     },
-    "outputRowNumber" : true,
+    "outputRowNumber" : false,
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-rank_1/rank-test/plan/rank-test.json:68]: Let's see...  

For RankProcessStrategy, the existing/ported tests only cover AppendFast.  (There are Undefined, Retract, and UpdateFast strategies.)
(I'm guessing that `generateUpdateBefore` is coupled with the strategy?)

For RankType, only ROW_NUMBER is covered.  There is RANK, and DENSE_RANK.  

For RankRange, we are testing having 1 and 2 rows...  seems like that is some reasonable coverage.  

What is useful here?  Adding coverage for the two (or three, if Undefined is possible) `RankProcessStrategy`s and testing the other RankTypes?

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-rank_1/rank-test/plan/rank-test.json:68]: > What is useful here? Adding coverage for the two (or three, if Undefined is possible) RankProcessStrategys and testing the other RankTypes?

This sounds reasonable, yes

[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/resources/restore-tests/stream-exec-rank_1/rank-test/plan/rank-test.json]: As an update, RANK, and DENSE_RANK are not supported for streaming; I missed that when reading initially.  

The "Undefined" strategy is set before an actual strategy has been assigned.  I have added tests for the other two.

Let me know if you find anything else to cover and/or have any naming suggestions.

**Comment by jnh5y** [comment]
Lemme know if you want me to clean up any of the git history.  The merge from master was resolving conflicts in the GitHub UI.

**Review by dawidwys — APPROVED** [review]
lgtm

Yes, it would be really nice if you could rebase it on top of current master. My personal take is we should never merge master into PRs. It makes reasoning about what has changed really hard. Moreover, I don't know how will that work with squashing, rebasing and so on. I see github's doing some magic to exclude the merged changes, but I don't know how that works.

**Pull request merged**